### ISSUE TITLE
[docu] qhelpgenerator on doxygen

### ISSUE THREAD

**Issue by ferdymercury** [root-project/root#8781]
### Explain what you would like to see improved
I daily use QtCreator in my ROOT-based projects, and it would be great if there would be a config-flag to build ROOT with the `GENERATE_QHP` enabled in the doxygen file. https://github.com/root-project/root/blob/master/documentation/doxygen/Doxyfile#L1514

This way, one can consult the Doxygen documentation directly within QtCreator, no need to use a web browser, which is also great if you don't have internet connection.
See e.g. https://bugreports.qt.io/secure/attachment/48119/Added_doxygen_doc_2.PNG

### Optional: share how it could be improved
cmake config flag, or automatically detect if qhelpgenerator is found as done in llvm
https://github.com/root-project/root/search?q=qhelpgenerator

### Setup
```
   ------------------------------------------------------------------
  | Welcome to ROOT 6.25/01                        https://root.cern |
  | (c) 1995-2021, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on May 17 2021, 10:49:42                 |
  | From heads/master@v6-25-01-861-g5ea261143f                       |
  | With                                                             |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------
```

**Comment by couet** [comment]
I am not completely sure how this "additional index file" can be used. Should it be made available for download from this page ? https://root.cern/reference/

**Comment by ferdymercury** [comment]
In my opinion, it does not matter too much where it is stored, as you have to go to QtCreator and add it to the "help database" by hand using a file dialog. Maybe including it in the binaries inside a documentation folder would be enough, so that one does not need to reupdate the webpage too often?

In any case, I think it would be great if it is automatically generated when building from source, always in the same position, so that one does not need to update the qtcreator database path pointer.

**Comment by couet** [comment]
So what you need is `GENERATE_QHP` enabled in Doxyfile while generating the reference guide , right ? Have you tried it ? does the result doxygen produces is what you need ?

**Comment by ferdymercury** [comment]
Thanks. Yes, that's what would be needed. I didn't try it with ROOT but it works well with all my other Cmake projects. It generates a .qch file that one can then import as Help Guide within QtCreator
See https://www.creatis.insa-lyon.fr/~grenier/?p=273

**Comment by couet** [comment]
Sorry for my ignorance but this .qch file contains what ? I guess it cannot be the whole reference guide ? that would be a huge file ... Is it references to the guide on the web ? in that case I do not understand your statement in your first post when you said: _which is also great if you don't have internet connection_

**Comment by ferdymercury** [comment]
See also https://reviews.llvm.org/D2967

**Comment by ferdymercury** [comment]
It is a compressed file qt help file, stored locally. Yes, it is potentially a very large file. (Could be enabled via cmake flag).

**Comment by couet** [comment]
We do not use cmake to build the reference guide. It is a separated Makefile.
Yes that file might be huge. How are managed the pictures ? they are somehow included in it ?
I will try what is suggested in your insa-lyon link

**Comment by couet** [comment]
I see the command `qhelpgenerator` needs to be installed. I do not have Qt (and do not want it) on my machine. Can `qhelpgenerator` be installed without the full Qt (which requires a licence if I remember well) ?

**Comment by ferdymercury** [comment]
You don't need a license for Qt if you click on OpenSource. The option is a bit hidden. Also, you can install just QtCreator without installing full Qt.
Yes, i think qhelpgenerator can be installed independently, check the LLVM link to see related apt packages.

**Comment by ferdymercury** [comment]
https://www.qt.io/download-open-source and when installing unselect all except for qtcreator. There are also offline installers that don't require login.

**Comment by ferdymercury** [comment]
Here the file format description https://doc.qt.io/qt-5/qthelp-framework.html
I think images are linked, not embedded https://forum.qt.io/topic/73965/linked-png-in-qthelp-file

**Comment by couet** [comment]
Ok, the installed went wrong. I re-did it and i see `qhelpgenerator` now.
I will test on a subset of the doc.

**Comment by couet** [comment]
I have modified the following entries in Doxyfile:
```
GENERATE_QHP = YES       
QCH_FILE  = ROOT.qch    
QHP_NAMESPACE = ROOT.Project
QHG_LOCATION = /Users/couet/Qt/6.1.2/macos/bin/qhelpgenerator
```
I regenerated the hist part of the reference guide on my local machine.
I was expecting the file `ROOT.qch` will appear somewhere . But it does not.

**Comment by ferdymercury** [comment]
Running doxygen with extra verbose flag might give you a warning message. It could be the path, see https://stackoverflow.com/a/49751322

**Comment by ferdymercury** [comment]
By the way, once you open qtcreator, you can click on File, open file or project, and double click on the main Cmakelists.txt

**Comment by couet** [comment]
It seems doxygen tried to run this `qhelpgenerator` but failed:
```
Running qhelpgenerator...
Error in line 15: Opening and ending tag mismatch.
error: failed to run qhelpgenerator on index.qhp
lookup cache used 13266/1048576 hits=365874 misses=14050
finished...
```
That's weird because very likely this index.php file has been generated by doxygen, and it seems to have a wrong syntax.

**Comment by ferdymercury** [comment]
Thanks for checking. What's the doxygen version? 

Could be
https://github.com/doxygen/doxygen/issues/5333
Or
https://github.com/doxygen/doxygen/issues/6254

**Comment by couet** [comment]
I install doxygen from sources . I am using the master version 1.9.2

**Comment by ferdymercury** [comment]
Thanks. Then it's probably due to https://github.com/doxygen/doxygen/issues/5333
Can you open index.qhp with a text editor and post here the content around line 15?
And maybe also in that open doxygen issue.

**Comment by ferdymercury** [comment]
How can I build the documentation to try myself? I build ROOT from source, but I do not see any doxygen output.

**Comment by couet** [comment]
we have a dedicated web page for developers [explaining exactly that](https://root.cern/for_developers/doxygen/).

**Comment by ferdymercury** [comment]
Thanks! Is there a page explaining the dependencies? I tried and I get this error:
`ImportError: No module named nbformat`

**Comment by couet** [comment]
Ah yes `nbformat` . That's use by `converttonotebook.py` to convert the tutorials to notebooks. May be I should add that. In principle that should not stop the build of the documentation . You are installing the tutorials ? you are trying to install the full doc ? (it is 5 hours long)

**Comment by ferdymercury** [comment]
Yes, I guess I was doing that hehe. If I want to build only the 'hist', how can I select it?

**Comment by ferdymercury** [comment]
`I installed in pip the nbconvert, but I still get: Exception: Jupyter command `jupyter-nbconvert` not found.`

**Comment by couet** [comment]
Yes you better do only a part of the manual. The way to do it is explained here. https://root.cern/for_developers/doxygen/#how-to-generate-the-doxygen-documentation

**Comment by ferdymercury** [comment]
Thanks. I get now this error:
`jupyter_client.kernelspec.NoSuchKernel: No such kernel named python2`

**Comment by ferdymercury** [comment]
Ok, I am now using just this as input:

`./mainpage.md ../../graf2d/gpad/`

It builds well, with just this warning:
```
/opt/root_src/documentation/doxygen/mainpage.md:4: warning: found subsubsection command outside of subsection context!
/opt/root_src/documentation/doxygen/mainpage.md:10: warning: found subsubsection command outside of subsection context!
/opt/root_src/documentation/doxygen/mainpage.md:18: warning: found subsubsection command outside of subsection context!
```

**Comment by ferdymercury** [comment]
Ok, I found the issue why qhelpgenerator is failing, it is due to:

```
/opt/root_src/documentation/doxygen/mainpage.md:4: warning: found subsubsection command outside of subsection context!
/opt/root_src/documentation/doxygen/mainpage.md:10: warning: found subsubsection command outside of subsection context!
/opt/root_src/documentation/doxygen/mainpage.md:18: warning: found subsubsection command outside of subsection context!
```

If I change in `mainpage.md` to:
```
\mainpage %ROOT Reference Documentation

# Introduction
Welcome to %ROOT!

This is the reference manual of the ROOT software tooklit.
You can find in the [reference documentation page](https://root.cern/reference/) pointers to reference manuals for all %ROOT versions.

## Other types of documentation:

- [ROOT Primer](https://root.cern/primer/).
- [ROOT Introductory Course](https://github.com/root-project/training/tree/master/BasicCourse).
- A rich set of %ROOT [tutorials and code examples](https://root.cern/doc/master/group__Tutorials.html) are offered to developers to exercise specific functionality.
- A general [Manual](https://root.cern/manual/) is provided for a more in depth explanation of concepts and functionality available in the %ROOT system.
- A number of topical [User Guides and Manuals](https://root.cern/topical/) for various components of the system.

## Provide your feedback
If you have suggestions about how to improve this documentation, you can let us know:

- With a [PR](https://github.com/root-project/root)
- On the [ROOT Forum](https://root-forum.cern.ch)
- On [our tracker](https://github.com/root-project/root/issues)
```
then `ROOT.qch` is correctly generated.

**Comment by couet** [comment]
What did you change in mainpage.md ?

**Comment by ferdymercury** [comment]
> What did you change in mainpage.md ?

The `###` to `#` or `##`. It complains that there is a subsubsection with no parent 'section'.

**Comment by couet** [comment]
Ah ok ... I do a PR for that ?
BTW I improved the Doxygen doc: https://github.com/root-project/web/pull/639

**Comment by couet** [comment]
> The `###` to `#` or `##`. It complains that there is a `subsubsection` with no parent 'section'.

Ah ok ... but that looks ugly (I tried) the headers are far too big. I guess we never use "#" only in the root ref guide. So fixing the main page might be not enough . The same problem might show with other pages. For instance here: https://github.com/root-project/root/blob/8a63f78a3f910b3bb8b7758a5af06a80d09e567f/hist/hist/src/TH1.cxx#L139

**Comment by ferdymercury** [comment]
The bug is gonna be solved soon in doxygen:
https://github.com/doxygen/doxygen/pull/8754

**Comment by ferdymercury** [comment]
I tried with building the docu of only the "core" libraries, and the qch file is correctly generated. When added to qtcreator, pressing F1 shows this beautifully:

![image](https://user-images.githubusercontent.com/10653970/131718506-8d759c57-a7d6-4314-8847-c80e7e37748a.png)

**Comment by ferdymercury** [comment]
> [](https://github.com/doxygen/doxygen/pull/8754)
> 
> In the meanwhile, we could compile doxygen using this fork/branch https://github.com/albert-github/doxygen/tree/feature/issue_8753

It's has now been merged into 'master', so no need to use a specific fork or branch, just git update.

**Comment by ferdymercury** [comment]
@couet the bug in doxygen is now fixed in official release doxygen version 1.9.3. https://github.com/doxygen/doxygen/issues/8753#issuecomment-1003347799

**Comment by couet** [comment]
@ferdymercury thanks to let me know. See the bottom of https://root.cern/doc/master/ . 1.9.3 is already the doxygen version used to build the ROOT reference guide. So we can close this issue ?

**Comment by ferdymercury** [comment]
> In my opinion, it does not matter too much where it is stored, as you have to go to QtCreator and add it to the "help database" by hand using a file dialog. Maybe including it in the binaries inside a documentation folder would be enough, so that one does not need to reupdate the webpage too often?
> 
> In any case, I think it would be great if it is automatically generated when building from source, always in the same position, so that one does not need to update the qtcreator database path pointer.

I would say it would be handier if the QCH is generated directly and is downloadable via the CERN website. (At least for stable releases 6.24, 6.26, ...).

**Comment by couet** [comment]
> For this to work, GENERATE_QHP should be set to True in the Doxyfiles of the ROOT repo, and then QHG_LOCATION set to a path where qhelpgenerator is installed in the ROOT build-servers (if not in the default path).

Can you propose a PR for that ?

**Comment by couet** [comment]
Thanks ! I would say the table on the page https://root.cern/reference/ should have one more column pointing to the qch. So I guess un extra PR is needed to modify https://github.com/root-project/web/blob/main/reference/index.md (may be an explanation saying what these files are and how to use them might be useful).

**Comment by ferdymercury** [comment]
I created it now. I just added the link for 6.24 as I do not think it worth to retrospectively add all the previous versions I guess...
https://github.com/root-project/web/pull/716

**Comment by couet** [comment]
Thanks . I merged. I guess we can close this issue now ?

**Comment by ferdymercury** [comment]
Thanks!

Mmm but the Qch file has to be uploaded first manually to https://root.cern/doc/v624/ROOT.qch (right now the link is broken).

Or is there a CI script that we have to modify so that this step happens automatically on every release ?

**Comment by couet** [comment]
I guess it should be copied automatically as ROOT.tag. The Scrip dubbing the doc are here:
https://github.com/root-project/rootspi/tree/master/rdoc
The script copying the doc at the right pace is https://github.com/root-project/rootspi/blob/master/rdoc/synchtml.sh . It copies the folder where the doc was generated in the right place to be accessed fro the web site. ROOT.qch should be like ROOT.tag

**Comment by ferdymercury** [comment]
Ok, I understand.
Then we are just missing the merge on https://github.com/root-project/root/pull/9525/files

Also: should we exclude ROOT.tag and ROOT.qch from the tarball in synchtml.sh ?

**Comment by couet** [comment]
> Also: should we exclude ROOT.tag and ROOT.qch from the tarball in synchtml.sh ?
Why ? it is something users might want in the tar-ball ?

**Comment by couet** [comment]
> Then we are just missing the merge 
done

**Comment by ferdymercury** [comment]
> > Also: should we exclude ROOT.tag and ROOT.qch from the tarball in synchtml.sh ?
> > Why ? it is something users might want in the tar-ball ?

Fair enough, I was just thinking to save some space, but you are right and it is probably useful to have it there, too.

**Comment by couet** [comment]
I think so. Disk space is not really an issue in that case.
I let you close this issue is you think all is done.

**Comment by ferdymercury** [comment]
> > Then we are just missing the merge
> > done

Thanks so much. Is there a way to resynch 6.24 so that the qch link on https://root.cern/reference/ is not broken ?

**Comment by couet** [comment]
I relaunched the build: https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-v624/

**Comment by ferdymercury** [comment]
Thanks so much.

Silly question, is qhelpgenerator installed in the build server?

If it's an old Ubuntu, one needs to:
- apt install qt4-dev-tools
- change in the Doxyfile: QHG_LOCATION=/usr/lib/x86_64-linux-gnu/qt4/bin/qhelpgenerator

In Ubuntu 20 or later, I think that it's
- apt install qhelpgenerator-qt5
- and no need to change the Doxyfile

**Comment by couet** [comment]
I am not sure to check the ubuntu version. `name -a` gives:
```
Linux root-ubuntu-2004-3 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
is that "old" or "new" ?

**Comment by ferdymercury** [comment]
Try with `cat /etc/lsb-release`

**Comment by couet** [comment]
ok "new":
```
DISTRIB_RELEASE=20.04
```

**Comment by ferdymercury** [comment]
Great! If you type `qhelpgenerator` in the command line, it should hopefully work. In Ubuntu 18, that command would give this error:
`qhelpgenerator: could not exec '/usr/lib/qt5/bin/qhelpgenerator': No such file or directory
`

**Comment by couet** [comment]
I get this exact error you posted

**Comment by couet** [comment]
Ok, I did the apt install as you suggested (with the admin account) and I now get:
```
$ qhelpgenerator
Missing input file name.


Usage:

qhelpgenerator <file> [options]

  -o <output-file>       Generates a Qt compressed help
                         called <output-file> (*.qch) for the
                         Qt help project <file> (*.qhp).
                         Generates a Qt help collection
                         called <output-file> (*.qhc) for the
                         Qt help collection project <file> (*.qhcp).
                         If this option is not specified
                         a default name will be used
                         (*.qch for *.qhp and *.qhc for *.qhcp).
  -c                     Checks whether all links in HTML files
                         point to files in this help project.
  -s                     Suppresses status messages.
  -v                     Displays the version of 
                         qhelpgenerator.

```
So I guess it is install ... should I relaunch the ref guide build ?

**Comment by ferdymercury** [comment]
Maybe check first the log file of the Doxygen build. Close to the end, it will say output an error like:
```
Running dot...
Patching output file 1/1
Running qhelpgenerator...
qhelpgenerator: could not exec '/usr/lib/qt5/bin/qhelpgenerator': No such file or directory
error: failed to run qhelpgenerator on index.qhp
lookup cache used 1000/65536 hits=4591 misses=1060
finished...
[100%] Built target doc
```

**Comment by couet** [comment]
Indeed the refi guide build for 6.24 did not start yet (looking for an executor) 
https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-v624/
so it should be fine.

**Comment by ferdymercury** [comment]
The build finished, but the link to the qch file is somehow still broken. Could you check the build log?

I tried the changes locally, and when I go to root_src/documentation/doxygen and call:
`doxygen`
it correctly creates the file ROOT.qch in the subfolder html.

```
/opt/root_src/documentation/doxygen$ ls html/*.qch -l
-rw-r--r-- 1 77799424 ene 11 16:48 html/ROOT.qch
```


Maybe the synch script is not copying correctly this created html file from the root source repository to ~/rootdoc ? Or maybe ROOT.tag is directly created in ~/rootdoc/html, whereas ROOT.qch is created in root_src/documentation/doxygen/html ?

Thanks again!

**Comment by couet** [comment]
Yes I saw that. Let me check.

**Comment by couet** [comment]
On this log https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-v624/96/consoleFull I search the string `qhelpgenerator ` but it does not appear. Should it ?

**Comment by ferdymercury** [comment]
Yes, I think so.
If I just call "doxygen" locally instead of make, it says at the end: 
```
Patching output file 1/1
Running qhelpgenerator...
```
And the Makefile is calling doxygen, so unless it's modifying the Doxyfile inbetween... it should.

**Comment by couet** [comment]
well: 
```
sftnight@root-ubuntu-2004-3:~/build/workspace/root-makedoc-v624/rootspi/rdoc/src/v6-24-00-patches/documentation/doxygen$ grep -i qch Doxyfile 
# (.qch) of the generated HTML documentation.
# If the QHG_LOCATION tag is specified, the QCH_FILE tag can be used to specify
# the file name of the resulting .qch file. The path specified is relative to
QCH_FILE               =
``` 
So it seems the doxyfile has not been update n the build machine

**Comment by couet** [comment]
Well .. the PR was on master ! that's why it does not show on 6.24

**Comment by ferdymercury** [comment]
ahh. good point, thanks for finding out. Can it be cherry picked into 6.24 or should I open a new PR ?

**Comment by couet** [comment]
Here (master) we see your mods: https://github.com/root-project/root/blob/master/documentation/doxygen/Doxyfile
But here (6.24) we don't: https://github.com/root-project/root/blob/v6-24-00-patches/documentation/doxygen/Doxyfile
but https://root.cern/doc/master/ROOT.qch does not work either

**Comment by ferdymercury** [comment]
Master ref guide was rebuilt before qhelpgenerator was installed maybe?

![image](https://user-images.githubusercontent.com/10653970/149124427-a8355b4c-2202-4da3-9f0e-cd220f3a1d60.png)

**Comment by couet** [comment]
let me check

**Comment by couet** [comment]
On the build machine Doxyfile seems ok for master:
```
sftnight@root-ubuntu-2004-3:~/build/workspace/root-makedoc-master/rootspi/rdoc/src/master/documentation/doxygen$ grep -i qch Doxyfile
# (.qch) of the generated HTML documentation.
# If the QHG_LOCATION tag is specified, the QCH_FILE tag can be used to specify
# the file name of the resulting .qch file. The path specified is relative to
QCH_FILE               = ROOT.qch
sft
```

**Comment by couet** [comment]
but may be that's the one going on ...

**Comment by ferdymercury** [comment]
I see also some "Kernel died" and "Traceback crashes", "python2 errros" in the log: https://lcgapp-services.cern.ch/root-jenkins/view/ROOT/job/root-makedoc-master/2580/consoleFull
```
 *** Break *** segmentation violation
Traceback...
```

but I guess this shouldn't be blocking the main doxygen process.

**Comment by couet** [comment]
No, it might be some tutorials failed.

**Comment by ferdymercury** [comment]
Alright, this looks better now in master, but there is still an error somewhere...

![image](https://user-images.githubusercontent.com/10653970/149142474-1e85146d-e8ea-488d-a109-d20deab8d3b9.png)

**Comment by couet** [comment]
Yes I see the command rans. The files are deleted once sync. Let me download the tar ball and see what's in.

**Comment by couet** [comment]
well there is no tarball for master... let me try to find out.

**Comment by couet** [comment]
I can see the files. The file ROOT.qch is missing:
```
$ ssh sftnight@root.cern.ch ls /var/www/root/doc/master/ROOT.*
/var/www/root/doc/master/ROOT.css
/var/www/root/doc/master/ROOT.tag
```

**Comment by couet** [comment]
here is [index.qhp.gz](https://github.com/root-project/root/files/7854927/index.qhp.gz)

**Comment by ferdymercury** [comment]
Yes, that's to be expected because of the error in the command:
`qhelpgenerator index.qhp`

I checked locally and I am seeing the same error. It seems the index.qhp file is corrupted here:
```

<toc>
      <section title="ROOT master" ref="index.html">
        <section title="ROOT Reference Documentation" ref="index.html">
          <section title="Introduction" ref="index.html#autotoc_md0" />
          <section title="Manuals" ref="index.html#autotoc_md1" />
          <section title="Tutorials and courses" ref="index.html#autotoc_md2" />
          <section title="Provide your feedback" ref="index.html#autotoc_md3" />
        </section>
      </section>
    </section>
    <section title="Tutorials" ref="group__Tutorials.html" />
```

**Comment by ferdymercury** [comment]
In the file you sent me, if you remove line 16, does the error disappear when running qhelpgenerator index.qhp ?

**Comment by ferdymercury** [comment]
It seems the error from https://github.com/doxygen/doxygen/issues/5333

**Comment by couet** [comment]
yes removing the line 16 make it works.
What do you suggest ?

**Comment by ferdymercury** [comment]
I'll report it, let's hope it gets fixed for 1.9.4

In the meanwhile, maybe we can document the "enums" in the TOC, so that the error does not appear?

**Comment by couet** [comment]
I can try to install the doxygen head ... that what we use already

**Comment by couet** [comment]
Let me reopen the issue until that's fixed.

**Comment by ferdymercury** [comment]
Yep, but it seems that this issue is still buggy in master. They suggest a workaround, but not sure...

```
QUICK FIXES FOR USERS:

If you personally encounter this "Error in line 4595: Opening and ending tag
mismatch.", and you just want to fix that issue, then the solution is to
either make sure that all your enums are documented or that they are
appropriately hidden (nested in an undocumented class, or nested in a
filtered-out namespace (e.g., "detail")). This is what I did, it fixed all
my issues.
```

**Comment by couet** [comment]
ok, I guess given the number of enums in ROOT the fix is not really applicable

**Comment by ferdymercury** [comment]
Yep, I agree.

Otherwise, the resulting qch after removing that line looks good when opened in QtCreator:

![image](https://user-images.githubusercontent.com/10653970/149161405-67891281-1f36-43f4-8283-3065d8a57961.png)

**Comment by couet** [comment]
We had our daily ROOT meeting this morning and It was suggested that you write a BlogPost on the ROOT web site to explain what is this qch file and how to use it. I think it is a good idea. Of course if you are willing.

**Comment by ferdymercury** [comment]
Sounds great, I would be happy to do it, thanks for the invitation.

But I would maybe like to wait that the doxygen bug gets fixed, before starting to write it.

**Comment by couet** [comment]
Yes may be better to have that doxygen fix before. The blog post can be prepared before as it is just a new PR on the ROOT web site. Having a PR allow the ROOT team to have a look at your text before posting. Let me know if you want more details about it.

**Comment by ferdymercury** [comment]
Doxygen has fixed the issue in Git master now.

**Comment by couet** [comment]
Ok. let me update my doxygen locally and on the build machine.

**Comment by couet** [comment]
I have updated doxygen to master on the build machine

**Comment by ferdymercury** [comment]
Thanks!

It seems we are getting now a different kind of error on the jenkins master log:
```
PatchinError in line 1: Premature end of document.
error: failed to run qhelpgenerator on index.qhp
g output file 29883/29885
Patching output file 29884/29885
Patching output file 29885/29885
Running qhelpgenerator...
lookup cache used 484195/1048576 hits=9429473 misses=518740
finished...
```

Could you share the index.qhp file to take a closer look? Maybe it's a multithreading issue?

**Comment by couet** [comment]
I sent it to you by email

**Comment by ferdymercury** [comment]
Thanks. I think I found a solution. Please try the following:

```
sed -i -e '/^ */s///' -e '/ *$/s///' index.qhp
qhelpgenerator index.qhp
```

**Comment by couet** [comment]
Seems ok:
```
$ qhelpgenerator index.qhp
Building up file structure...
Insert custom filters...
Insert help data for filter section (1 of 1)...
Insert files...
Warning: The file /home/sftnight/tabs.css does not exist, skipping it...
Warning: The file /home/sftnight/tab_a.png does not exist, skipping it...
Warning: The file /home/sftnight/tab_b.png does not exist, skipping it...
Warning: The file /home/sftnight/tab_h.png does not exist, skipping it...
Warning: The file /home/sftnight/tab_s.png does not exist, skipping it...
Warning: The file /home/sftnight/nav_h.png does not exist, skipping it...
Warning: The file /home/sftnight/nav_f.png does not exist, skipping it...
[...]
Warning: The file /home/sftnight/globals_defs_w.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_w.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_x.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_x.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_y.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_y.html does not exist, skipping it...
Warning: The file /home/sftnight/globals_defs_z.html does not exist, skipping it...
Insert contents...
Insert indices...
Documentation successfully generated.
```

**Comment by ferdymercury** [comment]
I filed https://github.com/doxygen/doxygen/issues/9058

In the meantime, I guess we could modify the Makefile script so that it does the following after doxygen finishes:
```
sed -i -e '/^ */s///' -e '/ *$/s///' index.qhp
qhelpgenerator index.qhp
```
and this part can be removed if the feature suggestion above is implemented in doxygen directly.

**Comment by couet** [comment]
ok. Let me see locally the exact command to add to the Makefile

**Comment by ferdymercury** [comment]
Thanks. (Side note: I took the command from here: https://stackoverflow.com/questions/59715196/how-can-i-efficiently-remove-unnecessary-whitespace-from-xml-files-with-a-linux)

**Comment by couet** [comment]
I added this line in the Makefile juste after the `doxygen` command:
```
	sed -i -e '/^ */s///' -e '/ *$/s///'  $(DOXYGEN_OUTPUT_DIRECTORY)/html/index.qhp
```

But what is executed is:

```
error: failed to run qhelpgenerator on index.qhp
lookup cache used 2073/1048576 hits=35556 misses=2201
finished...
sed -i -e '/^ */s///' -e '/ *s///'  /Users/couet/rootdoc/html/index.qhp
sed: -e expression #2, char 6: unknown command: `/'
make: *** [doxygen] Error 1
```

Somehow the `$` is not understood correctly when put in a Makefile. Do you know why ?
(I need to be away for my desk for a bit, let me know know if you know a solution).

**Comment by ferdymercury** [comment]
Not sure, maybe try a double dollar sign: https://til.hashrocket.com/posts/k3kjqxtppx-escape-dollar-sign-on-makefiles

**Comment by couet** [comment]
Yes double $ is ok. I'll make a PR

**Comment by couet** [comment]
PR here: https://github.com/root-project/root/pull/9649

**Comment by ferdymercury** [comment]
> I filed [doxygen/doxygen#9058](https://github.com/doxygen/doxygen/issues/9058)

The bug has been solved in current doxygen master.

**Comment by ferdymercury** [comment]
Hi @couet, https://github.com/doxygen/doxygen/issues/9094 has been fixed now in doxygen-master. Could you recompile it and give it another try? Thanks!!

**Comment by couet** [comment]
well: https://github.com/doxygen/doxygen/issues/9136

**Comment by couet** [comment]
I compile the new doxygen on the ubuntu build machine and ran it on the `root-docu ` you sent me while ago. I get:
```
[...]
finalizing index lists...
writing tag file...
Running plantuml with JAVA...
Running qhelpgenerator...
Building up file structure...
Insert custom filters...
Insert help data for filter section (1 of 1)...
Insert files...
Insert contents...
Insert indices...
Documentation successfully generated.
lookup cache used 11757/1048576 hits=120921 misses=12617
finished...

```
`Running qhelpgenerator...` seems to have passed ? right ?

**Comment by couet** [comment]
I still have the old doxygen on that machine and when I run it on `root-docu` I get:
```
Running qhelpgenerator...
Error in line 1: Opening and ending tag mismatch.
error: failed to run qhelpgenerator on index.qhp
lookup cache used 11757/1048576 hits=120921 misses=12617
finished...
```
So, yes, something has changed.

**Comment by ferdymercury** [comment]
Yes, cool, it seems to have been fixed. :)
Could you try to open the resulting .qch file with "assistant" (you'll need ssh -X) ? (To install, sudo apt install qt5-assistant, to run, just assistant). Then, in Preferences, Documentation, add the qch file in question.

**Comment by couet** [comment]
it seems ok:
<img width="1099" alt="Screenshot 2022-02-21 at 09 42 47" src="https://user-images.githubusercontent.com/4697738/154919012-25f28801-80d3-4c2b-9cac-06b03844802c.png">

So the only thing I need to do is to make this new doxygen the one used to build the doc ?

**Comment by ferdymercury** [comment]
Yes, that's right :)

**Comment by couet** [comment]
Done :-) lets wait the next build
Do you want to close this issue now ?
